### PR TITLE
Enable actions Zoom/Pan to Selection and Zoom to Layers. Fixes #40647

### LINKED ISSUE #40647: Pan Map to Selection in the View menu is enabled even for non-spatial and raster layers
The item "Pan Map to Selection" in the "View" menu is always enabled, even when a non-spatial layer or a raster is selected in the layer tree. A behavior like the "Zoom to Selection" would be expected where it's only enabled when a vector layers is selected in the layer tree.

Tested on 3.10 and also the current master branch from the git repository.

### PR DESCRIPTION
**Pull request by TurboGraphxBeige**
## Description

Enable or disable `mActionZoomToLayers`, `mActionZoomToSelected` and `mActionPanToSelected` actions according to the selected layers in the layer panel. It fixes the issue and represents better the new zooming and panning behavior using multiple layers.

The function `activateDeactivateMultipleLayersRelatedActions()` has been created to enable/disable the actions and is fired  
when changing the selection in the layer panel or when selected/deselecting features in any map layers.

`selectedLayersHaveSpatial()` and `selectedLayersHaveSelection()` functions have been created to check if any of the selected layers in the layer panel (or active layer if no selected layer) is spatial or has a selection.

Fixes https://github.com/qgis/QGIS/issues/40647

Peek 2021-01-31 18-43

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -14618,6 +14621,61 @@ void QgisApp::updateLabelToolButtons()
...
+  // If no selected layers, use active layer
+  if ( layers.size() == 0 )
+  {
+    QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( activeLayer() );
+
+    if ( layer && layer->selectedFeatureCount() == 0 )
+      return false;
+
+    if ( layer && layer->selectedFeatureCount() > 0 )
+      return true;
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:14637]:
```suggestion
    if ( QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( activeLayer() ) )
      return layer->selectedFeatureCount() > 0;
```

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -14618,6 +14621,61 @@ void QgisApp::updateLabelToolButtons()
...
+  }
+
+  return false;
+}
+
+void QgisApp::activateDeactivateMultipleLayersRelatedActions()
+{
+  mActionZoomToLayers->setEnabled( selectedLayersHaveSpatial() );
+  mActionZoomToSelected->setEnabled( selectedLayersHaveSelection() );
+  mActionPanToSelected->setEnabled( selectedLayersHaveSelection() );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:14676]:
```suggestion
  const bool hasSpatialSelection = selectedLayersHaveSpatial();
  mActionZoomToLayers->setEnabled( hasSpatialSelection );
  mActionZoomToSelected->setEnabled( hasSpatialSelection );
  mActionPanToSelected->setEnabled( hasSpatialSelection );
```

File: src/app/qgisapp.cpp (hunk 1 above)
Comment (TurboGraphxBeige) [reply src/app/qgisapp.cpp]: Yup!

File: src/app/qgisapp.cpp (hunk 2 above)
Comment (TurboGraphxBeige) [reply src/app/qgisapp.cpp]: Why only `const bool hasSpatialSelection = selectedLayersHaveSpatial();` but nothing for `selectedLayersHaveSelection`? See my 2nd commit https://github.com/qgis/QGIS/pull/41285/commits/9d56bf7d8002bf1210d6c3059186161ca7438c7e

Comment (nyalldawson) [reply src/app/qgisapp.cpp]: My mistake sorry!

**Comment by TurboGraphxBeige** [comment]
Thanks for reviewing and merging!

**Comment by agiudiceandrea** [comment]
Hi @TurboGraphxBeige @nyalldawson I tested on Windows the MXE build of this PR at 3f5f8ffa8c6b4eaa7f040df60e126b600c47f0cc and I found that QGIS crashes every time a layer is removed from the map if there is only one layer in the map. The QGIS MXE build at previous commit ffff5e1899286430882fbaa8443fe6ea549c4153 doesn't crash.

How to reproduce:
1. launch QGIS
2. create a new temporary scratch layer (e.g. point) -> a new empty map is displayed and the previous layer is added to the map
3. remove the layer or close the project or close QGIS -> QGIS crashes

File: src/app/qgisapp.cpp [hunk 3]
```diff
@@ -14618,6 +14621,59 @@ void QgisApp::updateLabelToolButtons()
...
 }
 
+bool QgisApp::selectedLayersHaveSelection()
+{
+  QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
+
+  // If no selected layers, use active layer
+  if ( layers.size() == 0 )
+  {
+    if ( QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( activeLayer() ) )
```
Comment (roya0045) [inline src/app/qgisapp.cpp:14631]: @agiudiceandrea Crash might be coming from here as there are no active layers.

**Comment by nyalldawson** [comment]
@agiudiceandrea fix is here: https://github.com/qgis/QGIS/pull/41319

**Comment by TurboGraphxBeige** [comment]
Thanks @agiudiceandrea for pointing that out and thanks @nyalldawson for the quick fix.

**Pull request merged**